docs/design/single-file-unsupported.md

```diff
@@ -0,0 +1,68 @@
+# Single-file unsupported attribute
+
+Starting with .NET Core 3.0 users have the ability to bundle all files used by their applications into a single executable through the publish functionality, `PublishSingleFile`. Although most applications can be agnostic of whether they will be deployed as single-file, there exist some APIs that are known to be incompatible with this deployment mode and thus can cause single-file apps to crash.
+
+While the existing single-file analyzer can help us emit warnings whenever incompatible runtime library APIs are called from within user's code in a project that has been marked with the intent of being published as single-file (through setting `PublishSingleFile` to true in its .csproj), developers can still take a dependency on third-party components implementing functionality that inadvertently make their application no longer compatible with single-file.
+
+## User Experience
+
+For library authors, the new attribute will allow them to easily annotate APIs known to be problematic when used in the context of a single-file application.
+
+```C#
+[SingleFileUnsupported("'Bar' method is not compatible with single-file", "https://help")]
+void Bar()
+{
+    Assembly executingAssembly = Assembly.GetExecutingAssembly();
+    var codeBase = executingAssembly.CodeBase;
+    ...
+}
+```
+
+Consumers of this library who want to publish their application as a single-file binary will commonly do it through setting the corresponding property in the application’s .csproj:
+
+```XML
+<PropertyGroup>
+    <PublishSingleFile>true</PublishSingleFile>
```

Comment:
This does not work for other single-file setups. We need a general way to define which capability link to SDK/msbuild targets. For example, browser-wasm will have this one and others always set (and cannot be unset), are we going to hardcode wasm-browser RID to linker?

Reply:
We could change the SDK to set PublishSingleFile=true for browser-wasm. Or we will need a new property. There's an existing discussion on basically this topic here: https://github.com/dotnet/sdk/issues/14562.

Reply:
> We could change the SDK to set PublishSingleFile=true for browser-wasm.

Wouldn't that also trigger real publish single work which we do differently for wasm?

Reply:
If it would - then I consider that to be a bug in the SDK. In short `dotnet publish -r browser-wasm -p:PublishSingleFile=true` should already either fail or "ignore" the single-file setting (since it already does that anyway). If that's not the case we should fix it. If we do go with `PublishSingleFile` as the setting to drive this, then the wasm SDK should "ignore" it (meaning it's set to true by default anyway).